app/controllers/payments_controller.rb

```diff
@@ -0,0 +1,66 @@
+class PaymentsController < ApplicationController
+  before_action :authenticate_user!, except: [:gift, :create_gift]
+  before_action :skip_signed_up_users, except: [:gift, :create_gift]
+  before_action :set_price, except: :create
+
+  def new
+    render layout: 'sign_up_flow'
+  end
+
+  def gift
+    @gift = Gift.new
+    render layout: 'sign_up_flow'
+  end
+
+  def create_gift
+    @gift = Gift.new(gift_params)
+    @gift.build_coupon
+    @payment = Payment.new(payment_params)
+    @payment.email = @gift.your_email
+    @payment.coupon = found_coupon?
+    @payment.gift = true
+    if @gift.valid? && @payment.valid?
+      ActiveRecord::Base.transaction do
+        @gift.save
+        @payment.save(validate: false)
```

Comment:
Yeah, wow, possibly the worst decision made by Julian during that weekend :)

On Mon, Jun 1, 2015 at 12:36 PM, Sunny Juneja <notifications@github.com>
wrote:

> In app/controllers/payments_controller.rb
> <https://github.com/JoyFoodly/joyful_12/pull/5#discussion_r31461437>:
>
> > +  def gift
> > +    @gift = Gift.new
> > +    render layout: 'sign_up_flow'
> > +  end
> > +
> > +  def create_gift
> > +    @gift = Gift.new(gift_params)
> > +    @gift.build_coupon
> > +    @payment = Payment.new(payment_params)
> > +    @payment.email = @gift.your_email
> > +    @payment.coupon = found_coupon?
> > +    @payment.gift = true
> > +    if @gift.valid? && @payment.valid?
> > +      ActiveRecord::Base.transaction do
> > +        @gift.save
> > +        @payment.save(validate: false)
>
> really need to refactor this. the issue is that validate calls stripe for
> a token. save and valid? both call validations. perhaps i can memoize the
> value.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JoyFoodly/joyful_12/pull/5/files#r31461437>.
>